packages/paymasters/contracts/PermitERC20UniswapV3Paymaster.sol

```diff
@@ -120,22 +174,29 @@ contract PermitERC20UniswapV3Paymaster is BasePaymaster, ERC2771Recipient {
     internal
     override
     returns (bytes memory context, bool revertOnRecipientRevert) {
-        (signature);
-        require(approvalData.length == 0, "approvalData: invalid length");
-        if (relayRequest.relayData.paymasterData.length != 0) {
-            require(relayRequest.relayData.paymasterData.length >= 4, "paymastaData: missing method sig");
+        (signature, approvalData);
+
+        // paymasterData must contain the token address, and optionally a a valid "permit" call on the token.
+        require(relayRequest.relayData.paymasterData.length >= 20, "must contain token address");
+        IERC20Metadata token = _getTokenFromPaymasterData(relayRequest.relayData.paymasterData);
+        require(isTokenSupported(token),"unsupported token");
+        if (relayRequest.relayData.paymasterData.length != 20) {
+            require(relayRequest.relayData.paymasterData.length >= 24, "must contain \"permit\" and token");
             require(
-                permitMethodSignature == GsnUtils.getMethodSig(relayRequest.relayData.paymasterData),
-                "paymasterData: wrong method sig");
+                permitMethodSignatures[token] == GsnUtils.getMethodSig(relayRequest.relayData.paymasterData),
+                "wrong \"permit\" method sig");
             // execute permit method for this token
             // solhint-disable-next-line avoid-low-level-calls
-            (bool success, bytes memory ret) = address(token).call(relayRequest.relayData.paymasterData);
+            (bool success, bytes memory ret) = address(token).call(relayRequest.relayData.paymasterData[:relayRequest.relayData.paymasterData.length - 20]);
             require(success, string(abi.encodePacked("permit call reverted:", string(ret))));
         }
-        uint256 tokenPreCharge = _calculatePreCharge(relayRequest, maxPossibleGas);
+
+        uint256 priceQuote = uint256(priceFeeds[token].latestAnswer());
```

Comment:
why bother with "priceDivisor" (both here and in postOp) ?
Seems like we can align "priceQuote" to 18 decimals here, and then the "convert" functions don't need to handle the divisor anymore.

Reply:
Because for some tokens it's not 18 and I wanted to keep it compatible